code/network/lib/Loot/Network/ZMQ/Adapter.hs

```diff
@@ -6,65 +6,99 @@
 -- ZMQ sockets.
 
 module Loot.Network.ZMQ.Adapter
-       ( atLeastOne
-       , threadWaitReadSTMLong
-       , socketWaitReadSTMLong
+       ( SocketAdapter
+       , newSocketAdapter
+       , adapterRelease
+       , adapterWait
+       , adapterTry
+
+       , atLeastOne
        , canReceive
        ) where
 
 import Control.Concurrent.Async as A
-import Control.Concurrent.STM.TMVar (newTMVarIO, putTMVar, readTMVar, swapTMVar, takeTMVar,
-                                     tryReadTMVar)
+import Control.Concurrent.STM.TVar (TVar, readTVar, writeTVar)
 import Control.Monad.STM (retry)
 import Data.List.NonEmpty as NE
 import qualified GHC.Conc.IO as G
 import System.Posix.Types (Fd)
 import qualified System.ZMQ4 as Z
 
--- | Given a set of STM actions, returns all that succeed if at least
--- one does.
-atLeastOne :: NonEmpty (STM (Maybe a)) -> STM (NonEmpty a)
-atLeastOne l = fmap catMaybes (sequence (NE.toList l)) >>= \case
-    [] -> retry
-    x:xs -> pure $ x :| xs
+-- | Internal socket adapter state.
+data SocketState
+    = SSNotInitialised -- ^ Worker hasn't yet reached blocking section.
+    | SSWaiting        -- ^ Currently waiting for messages from socket.
+    | SSCanRead        -- ^ Data can be read from socket.
+    | SSJustRead       -- ^ Data was just read from socket.
+    deriving (Eq)
+
+-- | Socket-related type that adapts it to STM-style polling.
+data SocketAdapter = SocketAdapter (TVar SocketState) (A.Async ())
 
 -- | Produces stm action corresponding to the action "lock until it is
 -- possible to read from the socket". It doesn't guarantee that there
 -- is actually any data to read from it.
-threadWaitReadSTMLong :: Fd -> IO (STM (), STM Bool, IO ())
+threadWaitReadSTMLong :: Fd -> IO (TVar SocketState, A.Async ())
 threadWaitReadSTMLong fd = do
     -- Bool value inside indicates who should take action now. If it's
     -- false, it's updater thread, if it's true it's main wait thread.
-    m <- newTMVarIO False
+    m <- newTVarIO SSNotInitialised
+
     -- Fork updater thread.
     t <- A.async $ forever $ do
-        -- Take the content of the TMVar if it's False, otherwise skip.
-        atomically $ do
-            v <- readTMVar m
-            if v == False
-            then void (takeTMVar m)
-            else retry
+        -- Section 1
+        atomically $ readTVar m >>= \case
+            SSWaiting -> error "threadWaitReadSTMLong: can't happen"
+            SSCanRead -> retry
+            _         -> writeTVar m SSWaiting
```

Comment:
Can we have `state --condition-> outcome` scheme for this state machine?

Also, you write `SSWaiting` into var, then on the next cycle if it is there you `error`. This construction seems brittle.

Reply:
I've added the explanation to the `SocketState` haddock.